Modules/_sqlite/connection.c

```diff
@@ -1839,17 +1839,30 @@ pysqlite_connection_exit_impl(pysqlite_Connection *self, PyObject *exc_type,
                               PyObject *exc_value, PyObject *exc_tb)
 /*[clinic end generated code: output=0705200e9321202a input=bd66f1532c9c54a7]*/
 {
-    const char* method_name;
+    int commit = 0;
     PyObject* result;
 
     if (exc_type == Py_None && exc_value == Py_None && exc_tb == Py_None) {
-        method_name = "commit";
+        commit = 1;
+        result = pysqlite_connection_commit_impl(self);
     } else {
-        method_name = "rollback";
+        result = pysqlite_connection_rollback_impl(self);
     }
 
-    result = PyObject_CallMethod((PyObject*)self, method_name, NULL);
     if (!result) {
+        if (commit) {
+            /* If commit failed, rollback and keep the existing exception. If
+             * rollback also fails, overwrite the exception. We use
+             * sqlite3_exec() to avoid accidentally clearing the current
+             * exception when rolling back. */
+            int rc;
+            Py_BEGIN_ALLOW_THREADS
+            rc = sqlite3_exec(self->db, "rollback;", NULL, NULL, NULL);
+            Py_END_ALLOW_THREADS
+            if (rc != SQLITE_OK && !PyErr_Occurred()) {
+                _pysqlite_seterror(self->db);
```

Comment:
Is it possible to chain commit and rollback exceptions?

Reply:
Hm, PEP 3134 seems to be obsolete, and PEP 490 was rejected, but it seems I can use `PyErr_NormalizeException` for this? :)

Reply:
Aha, I found `PyException_SetCause` and `PyException_SetContext`. PEP 3134 is apparently not dead.

Reply:
I tried implementing this in 98d3adaf5bc02ce0db2fd93246af9d5d4139a508. PTAL.

Reply:
We should probably chain exceptions in the first part of `__exit__` as well:

1. `__exit__` is called with an exception
2. rollback fails => chain exceptions